packages/fx-core/src/plugins/resource/keyvault/v3/index.ts

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { hooks } from "@feathersjs/hooks/lib";
+import {
+  AzureSolutionSettings,
+  err,
+  FxError,
+  Inputs,
+  ok,
+  Result,
+  v2,
+  v3,
+} from "@microsoft/teamsfx-api";
+import * as path from "path";
+import { Service } from "typedi";
+import { ArmTemplateResult } from "../../../../common/armInterface";
+import { Bicep } from "../../../../common/constants";
+import { generateBicepFromFile } from "../../../../common/tools";
+import { CommonErrorHandlerMW } from "../../../../core/middleware/CommonErrorHandlerMW";
+import { getTemplatesFolder } from "../../../../folder";
+import { AzureResourceKeyVault } from "../../../solution/fx-solution/question";
+import { BuiltInFeaturePluginNames } from "../../../solution/fx-solution/v3/constants";
+import { Constants } from "../constants";
+
+@Service(BuiltInFeaturePluginNames.keyVault)
+export class KeyVaultPluginV3 implements v3.FeaturePlugin {
+  name = BuiltInFeaturePluginNames.keyVault;
+  displayName = "Key Vault Plugin";
+  async pluginDependencies?(ctx: v2.Context, inputs: Inputs): Promise<Result<string[], FxError>> {
+    return ok([BuiltInFeaturePluginNames.identity]);
+  }
+
+  @hooks([CommonErrorHandlerMW({ telemetry: { component: BuiltInFeaturePluginNames.keyVault } })])
+  async generateResourceTemplate(
+    ctx: v3.ContextWithManifestProvider,
+    inputs: v2.InputsWithProjectPath
+  ): Promise<Result<v2.ResourceTemplate[], FxError>> {
+    const solutionSettings = ctx.projectSetting.solutionSettings as
+      | AzureSolutionSettings
+      | undefined;
+    const pluginCtx = { plugins: solutionSettings ? solutionSettings.activeResourcePlugins : [] };
+    const bicepTemplateDirectory = path.join(
+      getTemplatesFolder(),
+      "plugins",
+      "resource",
+      "keyvault",
+      "bicep"
+    );
+
+    const provisionModuleResult = path.join(
+      bicepTemplateDirectory,
+      Constants.provisionModuleTemplateFileName
+    );
+    const provisionOrchestration = await generateBicepFromFile(
+      path.join(bicepTemplateDirectory, Bicep.ProvisionFileName),
+      pluginCtx
+    );
+    const provisionModules = await generateBicepFromFile(provisionModuleResult, pluginCtx);
+    const result: ArmTemplateResult = {
+      Provision: {
+        Orchestration: provisionOrchestration,
+        Modules: { keyVault: provisionModules },
+      },
+      Reference: {
+        m365ClientSecretReference: Constants.KeyVaultBicep.m365ClientSecretReference,
+        botClientSecretReference: Constants.KeyVaultBicep.botClientSecretReference,
+      },
+    };
+    return ok([{ kind: "bicep", template: result }]);
+  }
+  @hooks([CommonErrorHandlerMW({ telemetry: { component: BuiltInFeaturePluginNames.keyVault } })])
+  async addFeature(
```

Comment:
What's 'addFeature' used for?

Reply:
addFeature means enable some feature by adding artifacts, which includes anything like scaffold, generateArmTemplate, update manifest